# FIFA World Cup 2026 — Winner Prediction
## Notebook 01 — Exploratory Data Analysis

The goal of this notebook is to explore and understand all four datasets before building any features or models.

### Objectives
- Load and inspect all datasets
- Understand the structure, shape, and data types of each
- Identify missing values and inconsistencies
- Extract initial insights that will inform feature engineering

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Display settings
pd.set_option('display.max_columns', None)
plt.style.use('ggplot')

In [3]:
df_matches = pd.read_csv("C:/Users/Usuario/Desktop/the bridge/github/wc2026-predictor/data/raw/wc_matches_1930_2022.csv", encoding='latin-1')
df_matches.head()

,Key Id,Tournament Id,tournament Name,Match Id,Match Name,Stage Name,Group Name,Group Stage,Knockout Stage,Replayed,Replay,Match Date,Match Time,Stadium Id,Stadium Name,City Name,Country Name,Home Team Id,Home Team Name,Home Team Code,Away Team Id,Away Team Name,Away Team Code,Score,Home Team Score,Away Team Score,Home Team Score Margin,Away Team Score Margin,Extra Time,Penalty Shootout,Score Penalties,Home Team Score Penalties,Away Team Score Penalties,Result,Home Team Win,Away Team Win,Draw
0,1,WC-1930,1930 FIFA World Cup,M-1930-01,France v Mexico,group stage,Group 1,1,0,0,0,7/13/1930,15:00,S-193,Estadio Pocitos,Montevideo,Uruguay,T-28,France,FRA,T-44,Mexico,MEX,4–1,4,1,3,-3,0,0,0-0,0,0,home team win,1,0,0
1,2,WC-1930,1930 FIFA World Cup,M-1930-02,United States v Belgium,group stage,Group 4,1,0,0,0,7/13/1930,15:00,S-192,Estadio Gran Parque Central,Montevideo,Uruguay,T-80,United States,USA,T-06,Belgium,BEL,3–0,3,0,3,-3,0,0,0-0,0,0,home team win,1,0,0
2,3,WC-1930,1930 FIFA World Cup,M-1930-03,Yugoslavia v Brazil,group stage,Group 2,1,0,0,0,7/14/1930,12:45,S-192,Estadio Gran Parque Central,Montevideo,Uruguay,T-84,Yugoslavia,YUG,T-09,Brazil,BRA,2–1,2,1,1,-1,0,0,0-0,0,0,home team win,1,0,0
3,4,WC-1930,1930 FIFA World Cup,M-1930-04,Romania v Peru,group stage,Group 3,1,0,0,0,7/14/1930,14:50,S-193,Estadio Pocitos,Montevideo,Uruguay,T-59,Romania,ROU,T-54,Peru,PER,3–1,3,1,2,-2,0,0,0-0,0,0,home team win,1,0,0
4,5,WC-1930,1930 FIFA World Cup,M-1930-05,Argentina v France,group stage,Group 1,1,0,0,0,7/15/1930,16:00,S-192,Estadio Gran Parque Central,Montevideo,Uruguay,T-03,Argentina,ARG,T-28,France,FRA,1–0,1,0,1,-1,0,0,0-0,0,0,home team win,1,0,0


In [6]:
df_matches.columns.tolist()

['Key Id',
 'Tournament Id',
 'tournament Name',
 'Match Id',
 'Match Name',
 'Stage Name',
 'Group Name',
 'Group Stage',
 'Knockout Stage',
 'Replayed',
 'Replay',
 'Match Date',
 'Match Time',
 'Stadium Id',
 'Stadium Name',
 'City Name',
 'Country Name',
 'Home Team Id',
 'Home Team Name',
 'Home Team Code',
 'Away Team Id',
 'Away Team Name',
 'Away Team Code',
 'Score',
 'Home Team Score',
 'Away Team Score',
 'Home Team Score Margin',
 'Away Team Score Margin',
 'Extra Time',
 'Penalty Shootout',
 'Score Penalties',
 'Home Team Score Penalties',
 'Away Team Score Penalties',
 'Result',
 'Home Team Win',
 'Away Team Win',
 'Draw']

In [4]:
df_matches.shape

(964, 37)

In [7]:
df_matches.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 964 entries, 0 to 963
Data columns (total 37 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Key Id                     964 non-null    int64 
 1   Tournament Id              964 non-null    object
 2   tournament Name            964 non-null    object
 3   Match Id                   964 non-null    object
 4   Match Name                 964 non-null    object
 5   Stage Name                 964 non-null    object
 6   Group Name                 964 non-null    object
 7   Group Stage                964 non-null    int64 
 8   Knockout Stage             964 non-null    int64 
 9   Replayed                   964 non-null    int64 
 10  Replay                     964 non-null    int64 
 11  Match Date                 964 non-null    object
 12  Match Time                 964 non-null    object
 13  Stadium Id                 964 non-null    object
 14  Stadium Na

In [8]:
df_matches['Stage Name'].value_counts()

Stage Name
group stage           676
round of 16            97
quarter-finals         70
semi-finals            38
second group stage     36
final                  21
third-place match      20
final round             6
Name: count, dtype: int64

In [9]:
df_matches['Result'].value_counts()

Result
home team win    545
away team win    240
draw             179
Name: count, dtype: int64

## Dataset 1 — Historical WC Matches (1930–2022)
- 964 rows, 37 columns
- Zero null values across all columns
- Encoding: latin-1 (special characters in team/city names)
- `Match Date` stored as object — needs datetime conversion in feature engineering
- `Score Penalties` stored as string — needs parsing in feature engineering
- Stage distribution: group stage dominates (676 matches), second group stage and final round are legacy formats to treat as group stage
- Result distribution: home team wins (545), away team wins (240), draws (179) — class imbalance to handle in model training with class weights

In [ ]:
df_elo = pd.read_csv(r"C:\Users\Usuario\Desktop\the bridge\github\wc2026-predictor\data\raw\elo_ratings_wc2026.csv")
df_train = pd.read_csv(r"C:\Users\Usuario\Desktop\the bridge\github\wc2026-predictor\data\raw\team_features_train.csv")
df_test = pd.read_csv(r"C:\Users\Usuario\Desktop\the bridge\github\wc2026-predictor\data\raw\team_features_test.csv")
df_matches_2026 = pd.read_csv(r"C:\Users\Usuario\Desktop\the bridge\github\wc2026-predictor\data\raw\wc2026_matches.csv")
df_teams_2026 = pd.read_csv(r"C:\Users\Usuario\Desktop\the bridge\github\wc2026-predictor\data\raw\wc2026_teams.csv")
df_stages_2026 = pd.read_csv(r"C:\Users\Usuario\Desktop\the bridge\github\wc2026-predictor\data\raw\wc2026_tournament_stages.csv")
df_host_cities = pd.read_csv(r"C:\Users\Usuario\Desktop\the bridge\github\wc2026-predictor\data\raw\wc2026_host_cities.csv")

In [12]:
for name, df in {
    "df_elo": df_elo,
    "df_train": df_train,
    "df_test": df_test,
    "df_matches_2026": df_matches_2026,
    "df_teams_2026": df_teams_2026,
    "df_stages_2026": df_stages_2026,
    "df_host_cities": df_host_cities
}.items():
    print(f"{name}: {df.shape}")

df_elo: (4683, 23)
df_train: (192, 24)
df_test: (48, 24)
df_matches_2026: (104, 8)
df_teams_2026: (48, 5)
df_stages_2026: (7, 3)
df_host_cities: (16, 6)


In [13]:
df_elo.columns.tolist()

['year',
 'snapshot_date',
 'country',
 'rank',
 'country_code',
 'rating',
 'rank_max',
 'rating_max',
 'rank_avg',
 'rating_avg',
 'rank_min',
 'rating_min',
 'matches_total',
 'matches_home',
 'matches_away',
 'matches_neutral',
 'wins',
 'losses',
 'draws',
 'goals_for',
 'goals_against',
 'confederation',
 'is_host']

In [14]:
df_elo['snapshot_date'].value_counts().head(10)

snapshot_date
1997-12-31    48
1998-12-31    48
1999-12-31    48
2000-12-31    48
2001-12-31    48
2002-12-31    48
2003-12-31    48
2004-12-31    48
2005-12-31    48
2006-12-31    48
Name: count, dtype: int64

In [16]:
df_elo['snapshot_date'].sort_values().unique()[-10:]

array(['2018-12-31', '2019-12-31', '2020-12-31', '2021-12-31',
       '2022-12-31', '2023-12-31', '2024-12-31', '2025-12-31',
       '2026-05-27', '2026-12-31'], dtype=object)

In [17]:
df_elo_pretournament = df_elo[df_elo['snapshot_date'] == '2026-05-27']

## Dataset 2 — Elo Ratings (2026 WC Teams)
- 4,683 rows — 125 years of Elo snapshots across all 48 qualified teams, 23 columns
- One snapshot per team per year-end, plus a live pre-tournament snapshot on `2026-05-27`
- Pre-tournament snapshot to use: `2026-05-27` (updated 2 weeks before kickoff) — filters down to exactly 48 rows, one per team
- `2026-12-31` snapshot must be avoided — contains post-tournament data (data leakage)
- Key features: `rating`, `rank`, `confederation`, `is_host`
- Useful derived features: `rating_max`, `rating_avg`, `rating_min`, `wins`, `losses`, `draws`, `goals_for`, `goals_against`, `matches_total`
- Columns to drop: `year`, `snapshot_date`, `country_code`, `matches_home`, `matches_away`, `matches_neutral`

In [18]:
df_train.columns.tolist()

['version',
 'team',
 'continent',
 'is_host',
 'goals_scored_last_4y',
 'goals_received_last_4y',
 'wins_last_4y',
 'losses_last_4y',
 'draws_last_4y',
 'world_cup_titles_before',
 'squad_total_market_value_eur',
 'fifa_rank_pre_tournament',
 'fifa_points_pre_tournament',
 'squad_avg_age',
 'world_cup_participations_before',
 'groups_passed_before',
 'round16_before',
 'quarterfinals_before',
 'semifinals_before',
 'finals_before',
 'winner',
 'finalist',
 'semi_finalist',
 'quarter_finalist']

In [21]:
df_train.head()

,version,team,continent,is_host,goals_scored_last_4y,goals_received_last_4y,wins_last_4y,losses_last_4y,draws_last_4y,world_cup_titles_before,squad_total_market_value_eur,fifa_rank_pre_tournament,fifa_points_pre_tournament,squad_avg_age,world_cup_participations_before,groups_passed_before,round16_before,quarterfinals_before,semifinals_before,finals_before,winner,finalist,semi_finalist,quarter_finalist
0,2006,Angola,Africa,0,61,49,19,13,14,0,52700000.0,57,581.0,29.2,0,0,0,0,0,0,0,0,0,0
1,2006,Argentina,South America,0,97,55,31,10,10,2,777200000.0,9,746.0,27.8,13,10,5,6,4,4,0,0,0,1
2,2006,Australia,Oceania,0,101,34,23,8,5,0,48030000.0,42,612.0,27.1,1,0,0,0,0,0,0,0,0,0
3,2006,Brazil,South America,0,117,47,30,9,17,5,858500000.0,1,827.0,27.1,17,15,7,11,9,6,0,0,0,1
4,2006,Costa Rica,North America,0,89,84,26,25,11,0,18400000.0,26,683.0,24.7,2,1,1,0,0,0,0,0,0,0


In [22]:
df_train.isnull().sum()

version                             0
team                                0
continent                           0
is_host                             0
goals_scored_last_4y                0
goals_received_last_4y              0
wins_last_4y                        0
losses_last_4y                      0
draws_last_4y                       0
world_cup_titles_before             0
squad_total_market_value_eur       32
fifa_rank_pre_tournament            0
fifa_points_pre_tournament          0
squad_avg_age                       0
world_cup_participations_before     0
groups_passed_before                0
round16_before                      0
quarterfinals_before                0
semifinals_before                   0
finals_before                       0
winner                              0
finalist                            0
semi_finalist                       0
quarter_finalist                    0
dtype: int64

In [25]:
df_test.head()

,version,team,continent,is_host,goals_scored_last_4y,goals_received_last_4y,wins_last_4y,losses_last_4y,draws_last_4y,world_cup_titles_before,squad_total_market_value_eur,fifa_rank_pre_tournament,fifa_points_pre_tournament,squad_avg_age,world_cup_participations_before,groups_passed_before,round16_before,quarterfinals_before,semifinals_before,finals_before,winner,finalist,semi_finalist,quarter_finalist
0,2026,France,Europe,0,85,32,25,6,7,2,1290000000,1,1877.32,26.7,16,10,8,9,7,4,NaN,NaN,NaN,NaN
1,2026,Spain,Europe,0,104,32,29,2,8,1,1150000000,2,1876.40,27.2,16,11,9,6,2,1,NaN,NaN,NaN,NaN
2,2026,Argentina,South America,0,80,14,30,4,3,3,575000000,3,1874.81,27.9,18,15,10,10,6,6,NaN,NaN,NaN,NaN
3,2026,England,Europe,0,82,23,26,6,7,1,1300000000,4,1825.97,26.8,16,13,8,10,3,1,NaN,NaN,NaN,NaN
4,2026,Portugal,Europe,0,98,31,26,5,7,0,841000000,5,1763.83,27.1,8,5,4,3,2,0,NaN,NaN,NaN,NaN


## Dataset 3 — Team Features (Train & Test)

### Train
- 192 rows, 24 columns — one row per team per historical World Cup edition (2006–2022)
- `version` column represents the World Cup year
- Target variables: `winner`, `finalist`, `semi_finalist`, `quarter_finalist` — binary flags
- Strong predictive features: `fifa_rank_pre_tournament`, `fifa_points_pre_tournament`, `goals_scored_last_4y`, `goals_received_last_4y`, `wins_last_4y`, historical stage columns
- `squad_total_market_value_eur` has 32 nulls (~17%) — decision to drop or impute pending in notebook 02
- Column to drop: `version`

### Test
- 48 rows, 24 columns — one row per qualified team for 2026
- Same structure as train but target columns are all `NaN` — to be predicted by the model
- `squad_total_market_value_eur` has no nulls — market value available for all current teams
- Top 5 by FIFA ranking: France, Spain, Argentina, England, Portugal

In [26]:
df_matches_2026.head()

,id,match_number,home_team_id,away_team_id,city_id,stage_id,kickoff_at,match_label
0,1,1,1.0,2.0,15,1,2026-06-11 15:00:00-06,Group A
1,2,2,3.0,4.0,14,1,2026-06-11 22:00:00-06,Group A
2,3,3,5.0,6.0,12,1,2026-06-12 15:00:00-04,Group B
3,4,4,13.0,14.0,6,1,2026-06-12 21:00:00-07,Group D
4,5,5,7.0,8.0,10,1,2026-06-13 15:00:00-07,Group B


In [28]:
df_teams_2026.head()

,id,team_name,fifa_code,group_letter,is_placeholder
0,1,Mexico,MEX,A,False
1,2,South Africa,RSA,A,False
2,3,South Korea,KOR,A,False
3,4,Winner UEFA Playoff D,UEPD,A,True
4,5,Canada,CAN,B,False


In [29]:
df_teams_2026['is_placeholder'].value_counts()

is_placeholder
False    42
True      6
Name: count, dtype: int64

In [30]:
df_teams_2026[df_teams_2026['is_placeholder'] == True]

,id,team_name,fifa_code,group_letter,is_placeholder
3,4,Winner UEFA Playoff D,UEPD,A,True
5,6,Winner UEFA Playoff A,UEPA,B,True
15,16,Winner UEFA Playoff C,UEPC,D,True
22,23,Winner UEFA Playoff B,UEPB,F,True
34,35,Winner FIFA Playoff 2,FP02,I,True
41,42,Winner FIFA Playoff 1,FP01,K,True


In [31]:
replacements = {
    'Winner UEFA Playoff A': ('Bosnia and Herzegovina', 'BIH'),
    'Winner UEFA Playoff B': ('Sweden', 'SWE'),
    'Winner UEFA Playoff C': ('Türkiye', 'TUR'),
    'Winner UEFA Playoff D': ('Czechia', 'CZE'),
    'Winner FIFA Playoff 1': ('DR Congo', 'COD'),
    'Winner FIFA Playoff 2': ('Iraq', 'IRQ')
}

for placeholder, (name, code) in replacements.items():
    mask = df_teams_2026['team_name'] == placeholder
    df_teams_2026.loc[mask, 'team_name'] = name
    df_teams_2026.loc[mask, 'fifa_code'] = code
    df_teams_2026.loc[mask, 'is_placeholder'] = False

df_teams_2026[df_teams_2026['fifa_code'].isin(['BIH', 'SWE', 'TUR', 'CZE', 'COD', 'IRQ'])]

,id,team_name,fifa_code,group_letter,is_placeholder
3,4,Czechia,CZE,A,False
5,6,Bosnia and Herzegovina,BIH,B,False
15,16,Türkiye,TUR,D,False
22,23,Sweden,SWE,F,False
34,35,Iraq,IRQ,I,False
41,42,DR Congo,COD,K,False


In [32]:
df_stages_2026.head(10)

,id,stage_name,stage_order
0,1,Group Stage,1
1,2,Round of 32,2
2,3,Round of 16,3
3,4,Quarterfinals,4
4,5,Semifinals,5
5,6,Third Place Playoff,6
6,7,Final,7


In [33]:
df_host_cities.head()

,id,city_name,country,venue_name,region_cluster,airport_code
0,1,Atlanta,USA,Mercedes-Benz Stadium,East,ATL
1,2,Boston,USA,Gillette Stadium,East,BOS
2,3,Dallas,USA,AT&T Stadium,Central,DAL
3,4,Houston,USA,NRG Stadium,Central,IAH
4,5,Kansas City,USA,Arrowhead Stadium,Central,MCI


In [34]:
df_host_cities['country'].value_counts()

country
USA       11
Mexico     3
Canada     2
Name: count, dtype: int64

## Dataset 4 — 2026 WC Match Data

### df_matches_2026
- 104 rows, 8 columns — all fixtures for the 2026 World Cup
- `home_team_id` and `away_team_id` are floats — foreign keys linking to `df_teams_2026`
- `city_id` links to `df_host_cities` — useful for host continent advantage feature
- `stage_id` links to `df_stages_2026` — stage progression logic for simulation
- `kickoff_at` confirms tournament starts June 11, 2026
- `match_label` identifies group (A–L) and knockout stage

### df_teams_2026
- 48 rows, 5 columns — one row per qualified team
- 6 placeholders updated manually with confirmed qualifiers:
  - UEFA Playoff A → Bosnia and Herzegovina (BIH)
  - UEFA Playoff B → Sweden (SWE)
  - UEFA Playoff C → Türkiye (TUR)
  - UEFA Playoff D → Czechia (CZE)
  - FIFA Playoff 1 → DR Congo (COD)
  - FIFA Playoff 2 → Iraq (IRQ)
- `is_placeholder` is now False for all 48 teams

### df_stages_2026
- 7 rows, 3 columns — one row per tournament stage
- Stages in order: Group Stage → Round of 32 → Round of 16 → Quarterfinals → Semifinals → Third Place Playoff → Final
- `stage_order` drives the progression logic in the Monte Carlo simulation engine
- Third Place Playoff is kept — relevant for podium probabilities (1st, 2nd, 3rd)

### df_host_cities
- 16 rows, 6 columns — one row per host city
- 11 USA, 3 Mexico, 2 Canada
- `region_cluster` (East/Central/West) is the key feature — used for host continent advantage
- Columns to drop: `venue_name`, `airport_code`

In [38]:
# df_matches.to_csv("../data/processed/df_matches_clean.csv", index=False)
# df_elo.to_csv("../data/processed/df_elo_clean.csv", index=False)
# df_train.to_csv("../data/processed/df_train_clean.csv", index=False)
# df_test.to_csv("../data/processed/df_test_clean.csv", index=False)
# df_matches_2026.to_csv("../data/processed/df_matches_2026_clean.csv", index=False)
# df_teams_2026.to_csv("../data/processed/df_teams_2026_clean.csv", index=False)
# df_stages_2026.to_csv("../data/processed/df_stages_2026_clean.csv", index=False)
# df_host_cities.to_csv("../data/processed/df_host_cities_clean.csv", index=False)

# print("All dataframes saved successfully.")